src/test/java/org/mockbukkit/mockbukkit/damage/DamageSourceMockTest.java

```diff
@@ -208,16 +244,56 @@ void equals_and_hashCode_WhenSameInstance()
 	void equals_and_hashCode_WhenIdentical()
 	{
 		DamageSourceMock copy = new DamageSourceMock(damageType, causingEntity, directEntity, damageLocation);
+		assertNotSame(damageSourceMock, copy);
 		assertEquals(damageSourceMock, copy);
 		assertEquals(damageSourceMock.hashCode(), copy.hashCode());
 		assertEquals(copy, damageSourceMock);
 		assertEquals(copy.hashCode(), damageSourceMock.hashCode());
 	}
 
+	@ParameterizedTest
+	@CsvSource({
+			"0,0,0,1",
+			"0,0,1,0",
+			"0,0,1,0",
+			"0,1,0,0",
+			"0,1,0,0",
+			"0,1,1,0",
+			"0,1,1,0",
+			"0,1,1,1",
+			"1,0,0,0",
+			"1,0,0,1",
+			"1,0,1,0",
+			"1,0,1,0",
+			"1,1,0,0",
+			"1,1,0,0",
+			"1,1,1,0",
+			"1,1,1,0",
+			"1,1,1,1"
+	})
+	void equals_and_hashCode_WhenNotIdentical(int a, int b, int c, int d)
+	{
+		boolean changedDamageType = a == 1;
+		boolean firstNull = b == 1;
+		boolean secondNull = c == 1;
+		boolean thirdNull = d == 1;
+
+		DamageSourceMock copy = new DamageSourceMock(
+				changedDamageType ? DamageType.ARROW : damageType,
+				firstNull ? null : causingEntity,
+				secondNull ? null : directEntity,
+				thirdNull ? null : damageLocation
+		);
+		assertNotEquals(damageSourceMock, copy);
+		assertNotEquals(damageSourceMock.hashCode(), copy.hashCode());
+		assertNotEquals(copy, damageSourceMock);
+		assertNotEquals(copy.hashCode(), damageSourceMock.hashCode());
+	}
+
 	@Test
 	void equals_and_hashCode_WhenNotHavingSameType()
 	{
-		assertNotEquals(damageSourceMock, new Object());
+		assertNotEquals(new Object(), damageSourceMock);
```

Comment:
Again, order of operations matter here, warning is being dumb.
```suggestion
		assertNotEquals(damageSourceMock, new Object());
```